### PR TITLE
Simplify root node resolution for SLURM environment

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Simplifies the root node resolution in SLURMEnvironment. No functional changes, but newly added test cases were added to cover untested special cases. 

This was noticed in the discussion in #14896. 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

cc @awaelchli @borda @akihironitta

### CODE REVIEW

File: tests/tests_lite/plugins/environments/test_slurm_environment.py [hunk 1]
```diff
@@ -81,7 +81,17 @@ def test_attributes_from_environment_variables(caplog):
 
 @pytest.mark.parametrize(
     "slurm_node_list,expected",
-    [("alpha,beta,gamma", "alpha"), ("alpha beta gamma", "alpha"), ("1.2.3.[100-110]", "1.2.3.100")],
```
Comment (awaelchli) [inline tests/tests_lite/plugins/environments/test_slurm_environment.py:84]: Note new test cases added, old ones still there.

File: src/lightning_lite/plugins/environments/slurm_environment.py [hunk 2]
```diff
@@ -134,7 +128,17 @@ def local_rank(self) -> int:
     def node_rank(self) -> int:
         return int(os.environ["SLURM_NODEID"])
 
-    def resolve_root_node_address(self, root_node: str) -> str:
+    @staticmethod
+    def resolve_root_node_address(nodes: str) -> str:
+        """The node selection format in SLURM supports several formats.
```
Comment (awaelchli) [inline src/lightning_lite/plugins/environments/slurm_environment.py:133]: Adding this for clarity, for future code readers, to make sure what is supported. If you look on the internet for SLURM documentation regarding this, good luck lol.

File: tests/tests_lite/plugins/environments/test_slurm_environment.py [hunk 3]
```diff
@@ -81,7 +81,17 @@ def test_attributes_from_environment_variables(caplog):
...
+    [
+        ("127.0.0.1", "127.0.0.1"),
+        ("alpha", "alpha"),
+        ("alpha,beta,gamma", "alpha"),
+        ("alpha beta gamma", "alpha"),
+        ("1.2.3.[100-110]", "1.2.3.100"),
+        ("1.2.3.[089, 100-110]", "1.2.3.089"),
+        ("host[5-9,1]", "host5"),
+        ("alpha, host[5-9], gamma", "alpha"),
```
Comment (awaelchli) [inline tests/tests_lite/plugins/environments/test_slurm_environment.py:92]: For this particular case, I'm not sure whether it is meant to be supported, i.e., mixing comma separated list with ranges. Can't find any documentation regarding this on the SLURM side. In any case we support parsing it.

**Review by carmocca — APPROVED** [review]

File: src/lightning_lite/plugins/environments/slurm_environment.py [hunk 4]
```diff
@@ -134,7 +128,17 @@ def local_rank(self) -> int:
...
+
+        This function selects the first host name from
+
+        - a space-separated list of host names, e.g., 'host0 host1 host3' yields 'host0' as the root
+        - a comma-separated list of host names, e.g., 'host0,host1,host3' yields 'host0' as the root
+        - the range notation with brackets, e.g., 'host[5-9]' yields 'host5' as the root
+        """
+        root_node = nodes.split(" ")[0].split(",")[0]
```
Comment (senarvi) [inline src/lightning_lite/plugins/environments/slurm_environment.py:141]: If you prefer, this function can be made simpler with a single regex:

    nodes = re.sub("\[(.*?)[,-].*?\]", "\\1", nodes)  # Take the first node of every node range
    return nodes.split(" ")[0].split(",")[0]

**Review by senarvi — APPROVED** [review]

File: src/lightning_lite/plugins/environments/slurm_environment.py (hunk 4 above)
Comment (awaelchli) [reply src/lightning_lite/plugins/environments/slurm_environment.py]: Yeah that's awesome, I integrated your suggestion

**Review by justusschock — APPROVED** [review]

**Pull request merged**